wgengine/tsdns/tsdns.go

```diff
@@ -0,0 +1,262 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package tsdns provides a Resolver struct capable of resolving
+// domains on a Tailscale network.
+package tsdns
+
+import (
+	"encoding/binary"
+	"errors"
+	"strings"
+	"sync"
+
+	"github.com/tailscale/wireguard-go/device"
+	dns "golang.org/x/net/dns/dnsmessage"
+	"inet.af/netaddr"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/packet"
+)
+
+// maxResponseSize is the maximal size of a Tailscale DNS response
+// including headers and wireguard padding.
+const maxResponseSize = 512
+
+// ipOffset is the length of wireguard padding before the IP header.
+const ipOffset = device.MessageTransportHeaderSize
+
+var (
+	errMapNotSet      = errors.New("domain map not set")
+	errNoSuchDomain   = errors.New("domain does not exist")
+	errNotImplemented = errors.New("query type not implemented")
+	errNotOurName     = errors.New("not an *.ipn.dev domain")
+	errNotQuery       = errors.New("not a DNS query")
+)
+
+var (
+	// The default IP for a new resolver.
+	DefaultIP = packet.IP(binary.BigEndian.Uint32([]byte{100, 100, 100, 100}))
+	// The default port for a new resolver.
+	DefaultPort = uint16(53)
+)
+
+// Map is all the data Resolver needs to resolve DNS queries.
+type Map struct {
+	// DomainToIP is a mapping of Tailscale domains to their IP addresses.
+	// For example, monitoring.ipn.dev -> 100.64.0.1.
+	DomainToIP map[string]netaddr.IP
+}
+
+// Resolver is a DNS resolver for domain names of the form *.ipn.dev
+type Resolver struct {
+	logf logger.Logf
+
+	// ip is the IP on which the resolver is listening.
+	ip packet.IP
+	// port is the port on which the resolver is listening.
+	port uint16
+
+	// parser is a request parser that is reused by the resolver.
+	parser dns.Parser
+	// responseBuffer is a static buffer to avoid graticious allocations.
+	responseBuffer [maxResponseSize]byte
+
+	// mu guards the following fields from being updated while used.
+	mu sync.Mutex
+	// dnsMap is the map most recently received from the control server.
+	dnsMap *Map
+}
+
+// NewResolver constructs a resolver with default parameters.
+func NewResolver(logf logger.Logf) *Resolver {
```

Comment:
If there is no constructor that takes non-default IP/port, I would make DefaultIP and DefaultPort private for now. We can always export them if we need it in future.